### PR TITLE
Relayer proxy hub processfromroot

### PR DESCRIPTION
**Pull request by rhlsthrm**
## Type of change

- [ ] Docs change / dependency upgrade
- [ ] Configuration / tooling changes
- [ ] Refactoring
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires changes in customer code

## Related Issue(s)

Fixes

### CODE REVIEW

File: packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol:77]: what about importing from `Types.sol`

File: packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol:388]: wont this always fall through and revert? i.e. if sending from optimism, it wouldn't exit early and fall into this edgecase. we likely need to return early out of the other functions

**Review by LayneHaber — CHANGES_REQUESTED** [review]

File: packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol
Comment (rhlsthrm) [reply packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol:388]: OMG you're right. Need elses.

**Review by LayneHaber — CHANGES_REQUESTED** [review]
tests :(

**Comment by rhlsthrm** [comment]
Unit tests added.

**Review by liu-zhipeng — APPROVED** [review]

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:35]: you shouldn't need to import errors, just events.

you can access errors like:

`ProposedOwnable.ProposedOwnable__onlyOwner_notOwner.selector`

when you import the `ProposedOwnable` file

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:80]:
```suggestion
for (uint256 I = 1; I <= _hubConnectors.length; I++) {
  _hubConnectors[I] = address(i)
}
```

File: packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol:326]: should we move these chains into a `Constants` or `ChainIds` library?

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:134]: interesting I thought `vm.prank` had to come after the `vm.expectEmit`

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:188]: can the root values be fuzzed?

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:224]: can you explain the priority mechanism to me? im not following what this test is testing

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:254]: it seems like this should be a util? its used several times in the test fns

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:255]: util?

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:287]: see above re: me being confused about priority mechanism

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:444]: "fuzz me" - `_message`

**Review by LayneHaber — COMMENTED** [review]
leaving as comments for now -- want @0xGorilla to review before anything is merged

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (rhlsthrm) [reply packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:217]: We have a mechanism for prioritizing based on block number. If the keeper is not the prioritized keeper, it can not send while block number < `_autonolasPriority`

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (rhlsthrm) [reply packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:437]: But this test isn't fuzzed, nor does it seem like it needs to be.

File: packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol
Comment (rhlsthrm) [reply packages/deployments/contracts/contracts/core/connext/helpers/RelayerProxyHub.sol]: Done.

File: packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol
Comment (rhlsthrm) [reply packages/deployments/contracts/contracts_forge/core/connext/helpers/RelayerProxyHub.t.sol:189]: Done.

File: packages/deployments/contracts/contracts/core/connext/libraries/ChainIDs.sol
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/connext/libraries/ChainIDs.sol:4]: amazing

**Review by LayneHaber — APPROVED** [review]

**Comment by LayneHaber** [comment]
still want a review from someone on the keeper side cc @0xGorilla @0xng

**Comment by LayneHaber** [comment]
they reviewed (see discord), lgtm

**Review by liu-zhipeng — APPROVED** [review]

**Comment by rhlsthrm** [comment]
To roll this out we will need to coordinate around migrating the credits and informing the keepers.

**Review by NaughtyUnicorn24 — APPROVED** [review]

**Comment by preethamr** [comment]
Not valid anymore

**Pull request closed without merging**
